linkerd/proxy/transport/src/prefix.rs

```diff
@@ -0,0 +1,75 @@
+use crate::io;
+use futures::prelude::*;
+use linkerd2_error::Error;
+use linkerd2_stack::layer;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time;
+use tower::util::ServiceExt;
+use tracing::debug;
+
+#[derive(Copy, Clone)]
+pub struct Prefix<S> {
+    inner: S,
+    capacity: usize,
+    timeout: Duration,
+}
+
+#[derive(Debug)]
+pub struct ReadTimeout(());
+
+impl<S> Prefix<S> {
+    pub fn new(inner: S, capacity: usize, timeout: Duration) -> Self {
+        Self {
+            inner,
+            capacity,
+            timeout,
+        }
+    }
+
+    pub fn layer(
+        capacity: usize,
+        timeout: Duration,
+    ) -> impl layer::Layer<S, Service = Prefix<S>> + Clone {
+        layer::mk(move |inner| Self::new(inner, capacity, timeout))
+    }
+}
+
+impl<S, I> tower::Service<I> for Prefix<S>
+where
+    I: io::Peekable + Send + 'static,
+    S: tower::Service<io::PrefixedIo<I>, Response = ()> + Clone + Send + 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = ();
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<(), Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(().into()))
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        debug!(capacity = %self.capacity, "Buffering prefix");
+        let accept = self.inner.clone();
+        let peek = time::timeout(self.timeout, io.peek(self.capacity)).map_err(|_| ReadTimeout(()));
+        Box::pin(async move {
+            let io = peek.await??;
+            accept.oneshot(io).err_into::<Error>().await?;
+            Ok(())
+        })
+    }
+}
+
+impl std::fmt::Display for ReadTimeout {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Timed out while reading client stream")
```

Comment:
it might be good if this error message mentioned that we were trying to read the prefix? although, on the other hand...we just fall back on these, so they don't really get exposed to the user, so the error message doesn't _really_ matter...

Reply:
We don't actually fall back on these, so these manifest as connection errors. I'm not sure that the prefix part of it matters, though. It's really just a read timeout -- the client hasn't provided any data.